src/journal/src/Journal/Types.hs

```diff
@@ -19,20 +21,258 @@ module Journal.Types
+  writeIntOffArrayIx meta
+    (tERM_TAIL_COUNTERS_OFFSET + (sizeOf (8 :: Int64) * partitionIndex))
+    (packTail termId0 termOffset0)
+
+casRawTail :: ByteBuffer -> Int -> Int64 -> Int64 -> IO Bool
+casRawTail meta partitionIndex expectedRawTail newRawTail =
+  casIntArray meta
+    (tERM_TAIL_COUNTERS_OFFSET + (sizeOf (8 :: Int64) * partitionIndex))
+    (fromIntegral expectedRawTail) (fromIntegral newRawTail) -- XXX: 32-bit systems?
+
+initialiseTailWithTermId :: ByteBuffer -> Int -> Int32 -> IO ()
+initialiseTailWithTermId meta partitionIndex termId0 =
+  setRawTail meta termId0 0 partitionIndex
+
+activeTermCount :: ByteBuffer -> IO Int32
+activeTermCount meta = readIntOffArrayIx meta lOG_ACTIVE_TERM_COUNT_OFFSET
+
+setActiveTermCount :: ByteBuffer -> Int32 -> IO ()
+setActiveTermCount meta = writeIntOffArrayIx meta lOG_ACTIVE_TERM_COUNT_OFFSET
+
+casActiveTermCount :: ByteBuffer -> Int32 -> Int32 -> IO Bool
+casActiveTermCount meta expectedTermCount newTermCount =
+  casIntArray meta lOG_ACTIVE_TERM_COUNT_OFFSET expectedTermCount newTermCount
+
+initialTermId :: ByteBuffer -> IO Int32
+initialTermId meta = readIntOffArrayIx meta lOG_INITIAL_TERM_ID_OFFSET
+
+-- should never be changed?
+-- setInitialTermId :: ByteBuffer -> Int32 -> IO ()
+-- setInitialTermId meta = writeInt32OffArrayIx meta lOG_INITIAL_TERM_ID_OFFSET
+
+termLength :: ByteBuffer -> IO Int32
+termLength meta = readIntOffArrayIx meta lOG_TERM_LENGTH_OFFSET
+
+-- should never be changed?
+-- setTermLength :: ByteBuffer -> Int32 -> IO ()
+-- setTermLength meta = writeInt32OffArrayIx meta lOG_TERM_LENGTH_OFFSET
+
+pageSize :: ByteBuffer -> IO Int32
+pageSize meta = readIntOffArrayIx meta lOG_PAGE_SIZE_OFFSET
+
+-- | The number of bits to shift when multiplying or dividing by the term buffer
+-- length.
+positionBitsToShift :: Int32 -> Int
+positionBitsToShift termBufferLength =
+  case termBufferLength of
+    65536      {- 64   * 1024 -}        -> 16
+    131072     {- 128  * 1024 -}        -> 17
+    262144     {- 256  * 1024 -}        -> 18
+    524288     {- 512  * 1024 -}        -> 19
+    1048576    {- 1024 * 1024 -}        -> 20
+    2097152    {- 2    * 1024 * 1024 -} -> 21
+    4194304    {- 4    * 1024 * 1024 -} -> 22
+    8388608    {- 8    * 1024 * 1024 -} -> 23
+    16777216   {- 16   * 1024 * 1024 -} -> 24
+    33554432   {- 32   * 1024 * 1024 -} -> 25
+    67108864   {- 64   * 1024 * 1024 -} -> 26
+    134217728  {- 128  * 1024 * 1024 -} -> 27
+    268435456  {- 256  * 1024 * 1024 -} -> 28
+    536870912  {- 512  * 1024 * 1024 -} -> 29
+    1073741824 {- 1024 * 1024 * 1024 -} -> 30
+    _otherwise ->
+      error ("positionBitsToShift: invalid term buffer length: " ++
+             show (termBufferLength))
+
+------------------------------------------------------------------------
+
+-- | Rotate to the next partition in sequence for the current term id.
+nextPartitionIndex :: Int32 -> Int32
+nextPartitionIndex currentIndex =
```

Comment:
Should we have newtypes for `TermIndex` and `PartitionIndex`?

Reply:
Yeah, there's a lot of `Int`s around now... I guess we can try to introduce newtypes as we start using all these functions where it makes sense? (Having to `fromIntegral` between newtypes also a bit annoying and error prone.)